transports/tcp/src/lib.rs

```diff
@@ -106,7 +107,10 @@ impl PortReuse {
     fn register(&mut self, ip: IpAddr, port: Port) {
         if let PortReuse::Enabled { listen_addrs } = self {
             log::trace!("Registering for port reuse: {}:{}", ip, port);
-            listen_addrs.insert((ip, port));
+            listen_addrs
+                .write()
+                .expect("`register()` and `unregister()` never panic while holding the lock")
```

Comment:
If you remove Clone on GenTcpConfig what about GenDnsConfig<GenTcpConfig, ...> which also implements Clone ?

Reply:
I would advocate for removing `Clone` from `GenDnsConfig` as well, though I would prefer that to happen in a separate pull request.

Can `GenDnsConfig` be `Clone` even if it is wrapping ` GenTcpConfig` which is not `Clone`?

Yes, given that `GenDnsConfig` wraps the `GenTcpConfig` in an `Arc<Mutex<_>>`.

Hope I am not missing something @pj50.

Reply:
I tried removing Clone from GenDnsConfig, however https://github.com/libp2p/rust-libp2/blob/4aa84bf7cd78c8553294f4f352fce9e20b6fbd4e/src/lib.rs#L206 does not compile any more. I don't know how to fix it properly, so I leave it to you, and you're right I think it would be much better to fix it in another pull request.